File: 451-sort-characters-by-frequency/sort-characters-by-frequency.cpp
class Solution {
public:
    static vector<int> freq;

    static bool compare(char a, char b) {
        if (freq[a] == freq[b])
            return a < b;  // Optional: lexicographic order for same frequency
        return freq[a] > freq[b];  // Higher frequency comes first
    }

    string frequencySort(string s) {
        // Reset the frequency array
        fill(freq.begin(), freq.end(), 0);

        // Count character frequencies
        for (int i = 0; i < s.length(); i++) {
            freq[s[i]]++;
        }

        // Sort using our static comparator
        sort(s.begin(), s.end(), compare);

        return s;
    }
};

// Define the static freq vector outside the class
vector<int> Solution::freq(128, 0);

File: 451-sort-characters-by-frequency/sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            if (map.containsKey(ch)) {
                map.put(ch, map.get(ch) + 1);
            } else {
                map.put(ch, 1);
            }
        }

        List<Character> ls = new ArrayList<>(map.keySet());

        Collections.sort(ls, (a, b) -> map.get(b) - map.get(a));

        StringBuilder sb = new StringBuilder();

        for (Character ch : ls) {
            for (int i = 0; i < map.get(ch); i++) {
                sb.append(ch);
            }
        }

        return sb.toString();
    }
}